### PR TITLE
FIX(client): Differentiate plugins for different OS 

### PR DESCRIPTION
**Pull request by Krzmbrzl**
Using which library can be loaded on the current OS is not a sufficient
criterion for differentiating between plugin libraries for different OS
as there is some overlap in supported file formats (e.g. .so is
supported on Linux and macOS).

Instead the new approach relies on suffixes in the plugin library's name
in order to determine which one to load.

### Checks

- [x] My commits follow the [commit guidelines](https://github.com/mumble-voip/mumble/blob/master/COMMIT_GUIDELINES.md)

### CODE REVIEW

File: src/mumble/PluginInstaller.cpp [hunk 1]
```diff
@@ -194,6 +210,10 @@ QString PluginInstaller::getInstallDir() {
 	return Global::get().qdBasePath.absolutePath() + QLatin1String("/Plugins");
 }
 
+QLatin1String PluginInstaller::getPluginSuffix() {
+	return QLatin1String("_" MUMBLE_TARGET_OS "_" MUMBLE_TARGET_ARCH);
```
Comment (davidebeatrici) [inline src/mumble/PluginInstaller.cpp:214]:
```suggestion
	return QLatin1String("_" MUMBLE_TARGET_OS "." MUMBLE_TARGET_ARCH);
```

**Comment by davidebeatrici** [comment]
Unfortunately `TargetArch.cmake` is almost a decade old, meaning that it doesn't know about ARM64.

**Review by hbeni — COMMENTED** [review]
:bulb: Idea: Additional to the hardwired naming scheme, introduce some optional _manifest_ file, which is searched for first.
That file (xml?) can map the OS and architecture to the binary names searched afterwards.

The manifest approach also enables flexibility in the future because we can define additional shipping files in there. This also allows us to drop the requirement for the shipped ZIP to contain solely binary plugin files, and allows the developer to structure the release package however she likes, including shipping additional, non-directly mumble related files.
Additional, not-yet-envisioned features can be configured in the manifest file also, making it upward compatible.

File: src/mumble/PluginInstaller.cpp (hunk 1 above)
Comment (Krzmbrzl) [reply src/mumble/PluginInstaller.cpp:214]: Why a dot? The docs already state to use an underscore :thinking:

Comment (davidebeatrici) [reply src/mumble/PluginInstaller.cpp:214]: Ah, I see. It was just for consistency with the binaries we release.

**Comment by Krzmbrzl** [comment]
Hm yeah a manifest file does sound like an interesting idea indeed :thinking: :+1: 

However I think I would make it mandatory and ditch the naming scheme in favor of the manifest... Seems easier to do (also facilitates and cleans up the implementation on Mumble's side).

**Comment by Krzmbrzl** [comment]
> Unfortunately TargetArch.cmake is almost a decade old, meaning that it doesn't know about ARM64.

@davidebeatrici do you know the required preprocessor statements to check for ARM64? In that case we could easily extend the script to support that as well :point_up:

**Comment by hbeni** [comment]
> However I think I would make it mandatory and ditch the naming scheme in favor of the manifest... Seems easier to do (also facilitates and cleans up the implementation on Mumble's side).

Sounds legit. I just proposed it to be optional because I didn't want you to break things.
But thats perfectly fine for development, and in this particular cases I want to second you that enforcing it is a very good idea, so this starts as clean as possible. And if its easier to handle, even better :smile:

**Comment by hbeni** [comment]
Just so i know how to proceed with my changes: when do you think you will introduce the mandatory manifest file, roughly?

**Comment by Krzmbrzl** [comment]
It depends on how much time I'll fins to work on this. I think for now we can say that I will only be able to look into this the coning weekend. It will probably be the next thing I am working on though and it shouldn't be too complicated to implement.

Therefore I would estimate that the respective PR will come sometime this or next weekend 🤔

**Comment by hbeni** [comment]
Do you have already ideas for the manifest format?
If no, i could write up a short proposal. I would propose a simple XML file format, beause that's easy to parse and extend but can be technically very precise.

Something like this:

Fixed name the updater will look for (case-insensitive): `manifest.mumble_plugin.xml`
```XML
<?xml version="1.0" encoding="UTF-8"?>
<plugin>
  <bundle version="1.0.0"/>   <!-- referring to the spec version in docs/dev/plugins/Bundling.md -->
  <assets>
    <!-- Mapping os/arch to library file names to look for in the package -->
    <file os="win"   arch="x86_64">foo.dll</file>
    <file os="win"   arch="i386">foo32.dll</file>
    <file os="linux" arch="x86_64">foo.so</file>
    <file os="macos" arch="x86_64">foo.bundle</file>
  </assets>
</plugin>
```

**Comment by Krzmbrzl** [comment]
I am currently undecided between using XML and JSON. Mainly because I have used JSON in C++ before whereas I have never had to deal with XML.
On the other hand though XML seems to be what most applications use for something like this and therefore what users probably expect (including me).
I guess it will depend on whether I find a good XML library that works to my liking in C++ ^^

I like your proposition but I would slightly alter it to something like this (mostly semantic changes):
```xml
<?xml version="1.0" encoding="UTF-8"?>
<bundle version="1.0.0"> <!-- referring to the spec version in docs/dev/plugins/Bundling.md -->  
  <assets>
    <!-- Mapping os/arch to library file names to look for in the package -->
    <plugin os="win"   arch="x86_64">foo.dll</file>
    <plugin os="win"   arch="i386">foo32.dll</file>
    <plugin os="linux" arch="x86_64">foo.so</file>
    <plugin os="macos" arch="x86_64">foo.bundle</file>
  </assets>
</bundle>
```

**Comment by davidebeatrici** [comment]
It's worth noting that we're currently using JSON to represent versions on our website:

- https://mumble.info/client-version.json
- https://mumble.info/server-version.json

**Comment by hbeni** [comment]
When you came up with the stable proposal for the manifest name and file format, let me know: i will happily adapt my plugin so you get quick test case ⏩

**Comment by Krzmbrzl** [comment]
Superseded by #5117

**Pull request closed without merging**